python/ray/serve/router.py

```diff
@@ -261,14 +267,16 @@ async def remove_worker(self, backend_tag, replica_tag):
     async def set_traffic(self, endpoint, traffic_dict):
         logger.debug("Setting traffic for endpoint %s to %s", endpoint,
                      traffic_dict)
-        self.traffic[endpoint] = traffic_dict
-        await self.flush()
+        self.traffic[endpoint] = self.policy(traffic_dict,
+                                             **self.policy_kwargs)
+
```

Comment:
I think we need to grab the flush lock here? There could a race condition where we in the middle of flushing endpoint queue somewhere and the policy got replaced. So we end up flushing to old backends. But I guess this is fine since we are grabbing the lock immediately afterward.

Reply:
ok i think the issue occurs when other coroutines _with_ the flush lock will inspect the `self.traffic` and assume a consistent view, but here we can mutate it _without_ the lock.

Reply:
Good catch, put it under the lock.